### ISSUE TITLE
current scanpy v1.10.3 cannot import igraph dependencies for sc.tl.leiden clustering algo

### ISSUE THREAD

**Issue by billiegould** [scverse/scanpy#3381]
### Please make sure these conditions are met

- [x] I have checked that this issue has not already been reported.
- [x] I have confirmed this bug exists on the latest version of scanpy.
- [ ] (optional) I have confirmed this bug exists on the main branch of scanpy.

### What happened?

Hi. I have been attempting for a while to get the latest scanpy jupyter notebook basic preprocessing tutorial to run on MacOS (Sonoma 14.5).
(https://scanpy.readthedocs.io/en/stable/tutorials/basics/clustering.html)

Is there a current version of scanpy with igraph dependency that is known to work stably using jupyter notebooks? This would be useful for other current tutorial users on Mac I believe.

I am only able to get the current pip3 install of scanpy (v1.10.3) to import and run in a jupyter notebook (conda installs not working).
At the leiden clustering step (sc.tl.leiden function). The function returns 
ImportError: Please install the igraph package: `conda install -c conda-forge python-igraph` or `pip3 install igraph`.

I have tried to get both pip3 and conda installs of several igraph and python-igraph versions working, and I have not been able to directly import any into the jupyter notebook (tried latest versions 0.11.8, earlier versions 0.10.2 and v0.9.10). 

I have also tried running the notebook inside the current scanpy docker container without success in importing igraph (https://hub.docker.com/r/gcfntnu/scanpy which has igrpah v.0.11.5).

Oddly, I am able to easily load all versions of igraph that I tried at the python3 interpreter without any issue. 

My current jupyter version is
$ jupyter --version
IPython          : 8.5.0
ipykernel        : 6.16.0
ipywidgets       : 8.0.2
jupyter_client   : 7.4.2
jupyter_core     : 4.11.1
jupyter_server   : 1.21.0
jupyterlab       : not installed
nbclient         : 0.7.0
nbconvert        : 7.2.1
nbformat         : 5.7.0
notebook         : 6.5.1
qtconsole        : 5.3.2
traitlets        : 5.5.0

(I can't figure out why the jupyter lab install is not listed here, but notebooks start up easily, so maybe part of the problem?)
Let me know if any further info on my system and package versions and locations would be helpful here. I feel like I've tried everything I know how so far to get the leiden clustering basic tutorial working. 

Thanks. 

### Minimal code sample

```python
See exact jupyter notebook tutorial code posted at 
https://scanpy.readthedocs.io/en/stable/tutorials/basics/clustering.html

Error at tutorial cell block:
sc.tl.leiden(adata, flavor="igraph", n_iterations=2)
```

### Error output

```pytb
---------------------------------------------------------------------------
ImportError                               Traceback (most recent call last)
Cell In [22], line 15
      7 flavor="igraph"
      8 # ImportError: dlopen(/Users/Billie/opt/ . . .
      9 
     10 #flavor="leidenalg"
     11 # ImportError: Please install the leiden algorithm: `conda install -c conda-forge leidenalg` or `pip3 install leidenalg`.
     12 # Requirement already satisfied: leidenalg in /Users/Billie/opt/anaconda3/envs/scrna/lib/python3.12/site-packages (0.10.2)
     13 # Requirement already satisfied: igraph<0.12,>=0.10.0 in /Users/Billie/opt/anaconda3/envs/scrna/lib/python3.12/site-packages (from leidenalg)
---> 15 sc.tl.leiden(adata, flavor=flavor, n_iterations=2)
     16 # sets fields:
     17 # `adata.obs['leiden' | key_added]` : :class:`pandas.Series` (dtype ``category``)
     18 #     Array of dim (number of samples) that stores the subgroup id
     19 #     (``'0'``, ``'1'``, ...) for each cell.
     20 print(adata.obs['leiden'].value_counts(dropna=False))

File /opt/homebrew/lib/python3.10/site-packages/scanpy/tools/_leiden.py:124, in leiden(adata, resolution, restrict_to, random_state, key_added, adjacency, directed, use_weights, n_iterations, partition_type, neighbors_key, obsp, copy, flavor, **clustering_args)
    120 if flavor not in {"igraph", "leidenalg"}:
    121     raise ValueError(
    122         f"flavor must be either 'igraph' or 'leidenalg', but '{flavor}' was passed"
    123     )
--> 124 _utils.ensure_igraph()
    125 if flavor == "igraph":
    126     if directed:

File /opt/homebrew/lib/python3.10/site-packages/scanpy/_utils/__init__.py:86, in ensure_igraph()
     84 if importlib.util.find_spec("igraph"):
     85     return
---> 86 raise ImportError(
     87     "Please install the igraph package: "
     88     "`conda install -c conda-forge python-igraph` or "
     89     "`pip3 install igraph`."
     90 )

ImportError: Please install the igraph package: `conda install -c conda-forge python-igraph` or `pip3 install igraph`.
```

### Versions

<details>

```

```

</details>

**Comment by flying-sheep** [comment]
Hi, this is almost certainly a problem with your environment(s).

You didn’t fill out the “Versions” section, which would have told us if you actually installed scanpy together with igraph at all. I think you didn’t.

Your notebook kernel uses your global Python installation `/opt/homebrew/lib/python3.10` (as the path for scanpy in your traceback is `/opt/homebrew/lib/python3.10/site-packages/scanpy`)

You probably installed igraph into some other environment. Make sure you use the kernel environment. E.g. you can use `%pip install igraph` in a notebook. But I wouldn’t recommend using global environments for analysis code.

Look here to learn more: https://janakiev.com/blog/jupyter-virtual-envs/

I’m going to close this. Please reply if you have more questions.

**Comment by billiegould** [comment]
Hello and thank you very kindly for your help solving my igraph and
other jupyter import issues with scanpy!
I had followed the current recommended scanpy install instructions posted
online, but I was unfamiliar with the approach of kernel customization in
jupyter and was instead attempting to add specific package locations
manually to my PYTHONPATH to fix import problems.
I started with a fresh reinstall of miniconda, scanpy, python-igraph, and
leidenalg from conda-forge, and then used ipykernel to create and run a
kernel from the env. Now a range of package import issues I had encountered
(and tried many hours to solve manually) have disappeared.

Thank you very much for your advice and the very useful weblink on
ipykernels! I hope this issue on git might help any other users that come
by with questions on getting the tutorials working in jupyter. You might
consider putting a line in the tutorial install webpage about the need to
use a custom jupyter kernel to get  all the scanpy dependencies to be
accessible, a tip for other beginning users like myself.

Thank you again very much for your time and your help, I'm looking forward
to lots of future work with this software.

All the best,
Billie
